cdk/dcm_floodlight/transform.js

```diff
@@ -0,0 +1,136 @@
+function isValidFlag(key, value) {
+  if (["true", "1"].includes(value.toString())) {
+    return 1;
+  }
+  if (["false", "0"].includes(value.toString())) {
+    return 0;
+  }
+
+  throw new CustomError(
+    `[DCM Floodlight]:: ${key}: valid parameters are [1|true] or [0|false]`,
+    400
+  );
+}
+
+function trackPostMapper(input, mappedPayload, rudderContext) {
+  const { message, destination } = input;
+  const { advertiserId, conversionEvents } = destination.Config;
+  let { activityTag, groupTag } = destination.Config;
+
+  const baseEndpoint = "https://ad.doubleclick.net/ddm/activity/";
+
+  let event = getValueFromMessage(message, "event");
+  if (!event) {
+    throw new CustomError(
+      "[DCM Floodlight]:: event is required for track call",
+      400
+    );
+  }
+
+  let customFloodlightVariable;
+  let salesTag;
+  event = event.trim().toLowerCase();
+  // find conversion event
+  // some() stops execution if at least one condition is passed and returns bool
+  const conversionEventFound = conversionEvents.some(conversionEvent => {
+    if (conversionEvent.eventName.trim().toLowerCase() === event) {
+      activityTag = conversionEvent.floodlightActivityTag.trim();
+      groupTag = conversionEvent.floodlightGroupTag.trim();
+      salesTag = conversionEvent.salesTag;
+      customFloodlightVariable = conversionEvent.customVariables || [];
+      return true;
+    }
+    return false;
+  });
+
+  if (!conversionEventFound) {
+    throw new CustomError("[DCM Floodlight]:: Conversion event not found", 400);
+  }
+
+  customFloodlightVariable = transformCustomVariable(customFloodlightVariable);
+  mappedPayload = {
+    src: advertiserId,
+    cat: activityTag,
+    type: groupTag,
+    ...mappedPayload
+  };
+
+  if (!salesTag) {
+    // for counter tag
+    mappedPayload.ord = get(message, "messageId");
+    delete mappedPayload.qty;
+    delete mappedPayload.cost;
+  }
+
+  const integrationsObj = getIntegrationsObj(message, "dcm_floodlight");
+  if (isDefinedAndNotNull(integrationsObj.COPPA)) {
+    mappedPayload.tag_for_child_directed_treatment = isValidFlag(
+      "COPPA",
+      integrationsObj.COPPA
+    );
+  }
+
+  if (isDefinedAndNotNull(integrationsObj.GDPR)) {
+    mappedPayload.tfua = isValidFlag("GDPR", integrationsObj.GDPR);
+  }
+
+  if (isDefinedAndNotNull(integrationsObj.npa)) {
+    mappedPayload.npa = isValidFlag("npa", integrationsObj.npa);
+  }
+
+  mappedPayload.dc_lat = get(message, "context.device.adTrackingEnabled");
+  if (isDefinedAndNotNull(mappedPayload.dc_lat)) {
+    mappedPayload.dc_lat = isValidFlag("dc_lat", mappedPayload.dc_lat);
+  }
+
+  mappedPayload = removeUndefinedAndNullValues(mappedPayload);
+  customFloodlightVariable = removeUndefinedAndNullValues(
+    customFloodlightVariable
+  );
+
+  rudderContext.endpoint = appendProperties(baseEndpoint, mappedPayload);
+  rudderContext.endpoint = appendProperties(
+    rudderContext.endpoint,
+    customFloodlightVariable
+  ).slice(0, -1);
```

Comment:
```suggestion
  const dcmEndpoint = appendProperties(baseEndpoint, mappedPayload);
  dcmEndpoint = appendProperties(
    dcmEndpoint,
    customFloodlightVariable
  ).slice(0, -1);
  
  rudderContext.endpoint = dcmEndpoint;
```
Lets not pass reference of rudderContext rather use temp variables